### ISSUE TITLE
wallabag v2 does not comply content security policy

### ISSUE THREAD

**Issue by oupala** [wallabag/wallabag#3070]
### Issue details

wallabag v2 does not comply content security policy

### Environment

* wallabag 2.2.2
* How did you install wallabag? by downloading the package
* Last wallabag version that did not exhibit the issue (if applicable): I don't know
* php version: PHP 5.6.29
* OS: raspbian
* type of hosting (shared or dedicated): dedicated
* which storage system you choose at install: SQLite

### Steps to reproduce/test case

When loading any page, the browser console is displaying many warnings: html contains css (styling), js (onclick behaviour). These should be moved to separate files in order to get a strict content security policy, which improve the security level.

```
Content Security Policy: Les paramètres de la page ont empêché le chargement d’une ressource à self (« default-src https://servername »). Source: .annotator-filter *,.annotator-notice,.a....  config:1
Content Security Policy: Les paramètres de la page ont empêché le chargement d’une ressource à self (« default-src https://servername »). Source: onfocusin attribute on DIV element. config
Content Security Policy: Les paramètres de la page ont empêché le chargement d’une ressource à self (« default-src https://servername »). Source: display: none. config
Content Security Policy: Les paramètres de la page ont empêché le chargement d’une ressource à self (« default-src https://servername »). config
Content Security Policy: Les paramètres de la page ont empêché le chargement d’une ressource à self (« default-src https://servername »). Source: 
                                    con....  config:267
Content Security Policy: Les paramètres de la page ont empêché le chargement d’une ressource à self (« default-src https://servername »). Source: onclick attribute on A element. config
```

**Comment by j0k3r** [comment]
Which CSP are you talking about?

**Comment by oupala** [comment]
I mean that the best CSP would be the following:

```
Content-Security-Policy "default-src 'self'"
```

And with this CSP, wallabag generates a lot of warnings.

**Comment by tcitworld** [comment]
The annotation system needs js (and probably css) inline, and you'll probably need to allow all external pictures.

**Comment by oupala** [comment]
I agree that I don't think we can avoid loading external images.

But why do annotation system needs inline js and css?

**Comment by tcitworld** [comment]
> I agree that I don't think we can avoid loading external images.

app.wallabag.it uses a camo system to have all pictures through https and a local domain, so there's no need for it here.

> But why do annotation system needs inline js and css?

It's an issue with http://annotatorjs.org/

**Comment by oupala** [comment]
So if you're clear about CSP (and you're in the top-3 commiter of wallabag), I think that wallabag should either:

- manage the CSP itself (to use the best CSP on each page)
- write in the documentation the best CSP that should apply on the whole app website

**Comment by j0k3r** [comment]
I'm 100% sure wallabag needs to declare **itself** which CSP user should use.
So I'll edit your issue.

**Comment by oupala** [comment]
@j0k3r I agree with you, the best solution is that Wallabag declare its own CSP. This is the best solution as the CSP could be different depending on the page. Developers can tell what permissions are needed by each page (and some page might require more permissions than others, it would be a shame to lower the general security level if only some pages require additional permissions).

But I also think that sometime, admins want to manage CSP on their own, and document should tell what is the best CSP to set for wallabag.

I'm willing to assist you and to test the CSP in order to tell you if there is sill some warnings or not.

**Comment by Crocmagnon** [comment]
Hello! I found this issue looking for the CSP to apply on wallabag. Since this hasn't seen any movement since 2017, I propose the following:

1. As a first iteration, we update the documentation (and this issue) with a basic CSP admins can safely apply on their instance without breaking anything but without allowing too much either. We apply this CSP in the app.wallabag.it instance
3. After a defined testing period, we integrate this default CSP in wallabag itself.
4. Then we can start working on custom per-page CSP if needed.

This should allow us to provide a first security barrier without too much hassle. I'm willing to help determine the right CSP for wallabag if you need help.
But the current situation is not satisfactory as we don't have anything at all 🙂

**Comment by oupala** [comment]
Here is my Content Security POlicy:

> Content-Security-Policy "upgrade-insecure-requests; default-src 'self' 'unsafe-inline'; img-src 'self' https:; frame-ancestors 'none'

If @j0k3r, @tcitworld or @Crocmagnon (or anyone else) want to discuss this topic, I'm willing to test different configurations in order to guess which one is the best.

**Comment by oupala** [comment]
My last comment was not good, it is not a good practice to allow "unsafe-inline" as a default src.

**Comment by DonPicciotto** [comment]
@oupala sadly it seems your CSP rule break images on my instance. The @Crocmagnon one seems the only working, but I don't think it's the right solution security-wise. Anyone can suggest a better alternative?
Of course thanks for your answers @oupala and @Crocmagnon !

**Comment by oupala** [comment]
Here is my current CSP rule :

> Content-Security-Policy "upgrade-insecure-requests; base-uri 'self'; default-src 'self'; script-src 'self' 'unsafe-inline'; img-src 'self' https:; frame-ancestors 'none';

Images are only authorized if they come from an https server.

**Comment by DonPicciotto** [comment]
> Here is my current CSP rule :
> 
> > Content-Security-Policy "upgrade-insecure-requests; base-uri 'self'; default-src 'self'; script-src 'self' 'unsafe-inline'; img-src 'self' https:; frame-ancestors 'none'
> 
> Images are only authorized if they come from an https server.

Thanks, and sorry for my previous message, it was wrong and after a couple of tests I decide to delete it. Your solution is the same as mine and my problems solved with `script-src 'unsafe-inline'` rule. It's still not satisfactory for me, tho, since it seems a bad option security-wise, but I will further test it. Thanks again for your answer!

**Comment by oupala** [comment]
Allowing `'unsafe-inline'` is indeed a bad option from a security point of view. How ever, not allowing `'unsafe-inline'` prevent the browser from executing some required javascript code, which break the behavior of Wallabag.

Changing this aspect of Wallabag would require to raise Wallabag to a higher security level, and this require development in order to move inline javascript (embedded into html) code to separate javascript files (`something.js`). So if you want to take part on that, feel free to propose your contribution to the Wallabag's developers. I'm sure they will be glad to here from you!

**Comment by DonPicciotto** [comment]
> Allowing `'unsafe-inline'` is indeed a bad option from a security point of view. How ever, not allowing `'unsafe-inline'` prevent the browser from executing some required javascript code, which break the behavior of Wallabag.
> 
> Changing this aspect of Wallabag would require to raise Wallabag to a higher security level, and this require development in order to move inline javascript (embedded into html) code to separate javascript files (`something.js`). So if you want to take part on that, feel free to propose your contribution to the Wallabag's developers. I'm sure they will be glad to here from you!
> 
> Edit: I've just seen that you are talking about `style-src 'unsafe-inline'` while I am talking about `script-src 'self' 'unsafe-inline'`. I hope we are talking about the same thing.
I initially messed up my message again, sorry, but yeah, I understand what you are saying. Sadly I don't have the skills to actually contribute to wallabag, even if I'd love to. Nevertheless, thanks for your time and explanation!

**Comment by hydrargyrum** [comment]
Having a CSP which is not the most secure and can be perfected later is probably better than not having a CSP at all, right?

Is this one sufficient as a start?

>     Content-Security-Policy: upgrade-insecure-requests; base-uri 'self'; default-src 'self'; script-src 'self' 'unsafe-inline'; img-src 'self' https:; frame-ancestors 'none'

**Comment by oupala** [comment]
As your CSP rule is the same as mine, I would answer that *yes*, this is sufficient.

Were you expecting another answer?

**Comment by bcutter** [comment]
I found `add_header Content-Security-Policy "default-src 'self'; script-src 'self' 'unsafe-inline'; style-src 'self' 'unsafe-inline'; img-src * data: blob:; font-src 'self' data:; connect-src *; frame-src *; media-src * data: blob:; frame-ancestors 'none';" always;` working fine, giving (among other settings of course) an A rating on https://securityheaders.com. Unfortunately an A+ is impossible (achievable by not using the `unsafe-inline` CSP), because of https://github.com/wallabag/wallabag/issues/8288.

Is the `unsafe-inline` CSP really necessary for Wallabag? I checked https://doc.wallabag.org/admin/installation/virtualhosts/#configuration-on-nginx and it seems like CSP is not set at all, so... functionality over security? Found this issue here from 2017... oh dear...
